pkg/billing/client.go

```diff
@@ -35,16 +35,138 @@ func GetHostname() string {
 	return hostname
 }
 
-func NewClient(url string, c *http.Client) Client {
-	return Client{URL: fmt.Sprintf("%s/usage_events", url), httpc: c}
+type Client interface {
+	Send(ctx context.Context, payload []byte, traceID TraceID) error
+	LogFields() zap.Field
 }
 
 type TraceID string
 
-func (c Client) GenerateTraceID() TraceID {
+func GenerateTraceID() TraceID {
 	return TraceID(shortuuid.New())
 }
 
+type HTTPClient struct {
+	URL   string
+	httpc *http.Client
+}
+
+func NewHTTPClient(url string, c *http.Client) HTTPClient {
+	return HTTPClient{URL: fmt.Sprintf("%s/usage_events", url), httpc: c}
+}
+
+func (c HTTPClient) Send(ctx context.Context, payload []byte, traceID TraceID) error {
+	r, err := http.NewRequestWithContext(ctx, http.MethodPost, c.URL, bytes.NewReader(payload))
+	if err != nil {
+		return RequestError{Err: err}
+	}
+	r.Header.Set("content-type", "application/json")
+	r.Header.Set("x-trace-id", string(traceID))
+
+	resp, err := c.httpc.Do(r)
+	if err != nil {
+		return RequestError{Err: err}
+	}
+	defer resp.Body.Close()
+
+	// theoretically if wanted/needed, we should use an http handler that
+	// does the retrying, to avoid writing that logic here.
+	if resp.StatusCode != http.StatusOK {
+		return UnexpectedStatusCodeError{StatusCode: resp.StatusCode}
+	}
+
+	return nil
+}
+
+func (c HTTPClient) LogFields() zap.Field {
+	return zap.String("url", c.URL)
+}
+
+type S3ClientConfig struct {
+	Bucket         string `json:"bucket"`
+	Region         string `json:"region"`
+	PrefixInBucket string `json:"prefixInBucket"`
+	Endpoint       string `json:"endpoint"`
+}
+
+type S3Client struct {
+	cfg    S3ClientConfig
+	client *s3.Client
+	now    func() time.Time
+}
+
+func NewS3Client(cfg S3ClientConfig, now func() time.Time) (S3Client, error) {
+	s3Config, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(cfg.Region))
+
+	if err != nil {
+		return S3Client{}, err
+	}
+
+	client := s3.NewFromConfig(s3Config, func(o *s3.Options) {
+		o.BaseEndpoint = &cfg.Endpoint
+		o.UsePathStyle = true // required for minio
+	})
+
+	return S3Client{
+		cfg:    cfg,
+		client: client,
+		now:    now,
+	}, nil
+}
+
+func (c S3Client) key() string {
+	// Example: year=2021/month=01/day=26/hh:mm:ssZ_{uuid}.ndjson.gz
+	now := c.now()
+	id := shortuuid.New()
+
+	filename := fmt.Sprintf("year=%d/month=%02d/day=%02d/%s_%s.ndjson.gz",
+		now.Year(), now.Month(), now.Day(),
+		now.Format("15:04:05Z"),
+		id,
+	)
+	return fmt.Sprintf("%s/%s", c.cfg.PrefixInBucket, filename)
+}
+
+type s3LogFields struct {
+	S3Client
+}
+
+func (c s3LogFields) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddString("bucket", c.cfg.Bucket)
+	enc.AddString("prefixInBucket", c.cfg.PrefixInBucket)
+	return nil
+}
+
+func (c S3Client) LogFields() zap.Field {
+	return zap.Inline(s3LogFields{c})
+}
```

Comment:
Maybe https://pkg.go.dev/go.uber.org/zap@v1.27.0/zapcore#ObjectMarshalerFunc could make this simpler? (i.e. implementing it all inside `LogFields()` instead of introducing a one-off type for ti)

Reply:
Indeed it does, thanks! Fixed in d6659f1

Reply:
Nice!